### PR TITLE
Add support for gfortran to be used with clang

### PR DESCRIPTION
**Pull request by BarrySmith**
1) list gfortran as a fc and f77 compiler that can work with clang
2) allow compatible gfortran to ./spack compiler find with clang by matching version numbers

This is based on the discussions in

https://github.com/LLNL/spack/issues/237
https://github.com/dealii/dealii/wiki/deal.II-in-Spack#mixing-gcc-and-clang-on-osx

This is not a long term solution but something to get us through the next months until the compiler
infrastructure is reworked to allow mixing and matching for C/C++ and Fortran compilers

Funded-by: IDEAS
Project: IDEAS/xSDK
Time: 1.5 hours

### CODE REVIEW

File: lib/spack/spack/compilers/clang.py [hunk 1]
```diff
@@ -121,6 +122,28 @@ def _find_full_path(self, path):
...
 
+    @classmethod
+    def fc_version(cls, fc):
+        version = get_compiler_version(
+            fc, '-dumpversion',
+            # older gfortran versions don't have simple dumpversion output.
+            r'(?:GNU Fortran \(GCC\))?(\d+\.\d+(?:\.\d+)?)')
+        # This is horribly ad hoc, we need to map from gcc/gfortran version
+        # to clang version, but there could be multiple clang
+        # versions that work for a single gcc/gfortran version
```
Comment (davydden) [inline lib/spack/spack/compilers/clang.py]: it's probably even anything to anything (I never had to worry which version of clang i use with what version of gfortran). But i don't have a better solution to this problem either.

**Comment by adamjstewart** [comment]
I am very much in favor of this change (assuming everything works fine, I haven't actually tested it). We shouldn't require users to hack Spack just to use Clang. And I think it's pretty safe to assume gfortran as a Fortran compiler on macOS. The only thing I would add is removing the hacking instructions from the documentation:
http://spack.readthedocs.io/en/latest/getting_started.html#mixed-toolchains

**Comment by tgamblin** [comment]
I'm totally in favor.

One question -- would it make sense to make a separate compiler called `clang-gfortran`, and allow *that* to be detected?  That would allow us to add `clang-pgi`, `clang-xl` etc., which are combinations that people use on other platforms.

**Comment by davydden** [comment]
>  would it make sense to make a separate compiler called clang-gfortran

would that break things like `spec.satisfies('%clang')`?

**Comment by tgamblin** [comment]
@davydden: It would.  That is a good reason *not* to do it and to use this solution as-is as a stopgap.

**Comment by citibeth** [comment]
Assuming you start with a bare Mac, we'll still need instructions on how to bootstrap: first build GCC, then somehow detect the clang+gfortran compiler.

**Comment by BarrySmith** [comment]
> Assuming you start with a bare Mac, we'll still need instructions on how to bootstrap: first build GCC, then somehow detect the clang+gfortran compiler.

First install homebrew http://brew.sh then just run

`brew install gcc`

then spack with my PR will find the gfortran and use it with clang.

**Comment by citibeth** [comment]
Isn't Spack capable of building GCC?  Why do we need to use Homebrew for
that?

**Comment by BarrySmith** [comment]
> Isn't Spack capable of building GCC?  Why do we need to use Homebrew for
that?

Yes presumably Spack could build gcc (using the clang compilers :-). Once you have built the gcc compilers you would need to run spack compiler find to get the combination of clang+gfortran and set that as the default.  As a user of spack, but not a developer, I don't really care so long as in the end I can use clang+gfortran*

Currently gcc cannot build packages such as openmpi on Mac https://github.com/LLNL/spack/issues/1847 so I need to use clang+gfortran to build a useable environment. Hence my interest in clang+gfortran.

**Comment by davydden** [comment]
@tgamblin supposedly, this is good to merge?

**Comment by tgamblin** [comment]
@davydden: yes -- can someone update the getting started guide?

**Comment by davydden** [comment]
@tgamblin see https://github.com/LLNL/spack/pull/2217

**Pull request merged**